python/cuml/preprocessing/TargetEncoder.py

```diff
@@ -0,0 +1,360 @@
+            return (cp.arange(len_train) /
+                    (len_train/self.n_folds)) % self.n_folds
+        elif self.split == 'interleaved':
+            return cp.arange(len_train) % self.n_folds
+        else:
+            msg = ("split should be either 'random'"
+                   " or 'continuous' or 'interleaved', "
+                   "got {0}.".format(self.split))
+            raise ValueError(msg)
+
+    def _compute_output(self, df_sum, df_count, cols, y_col):
+        """
+        Compute the output encoding based on aggregated sum and count
+        """
+        df_sum = df_sum.merge(df_count, on=cols, how='left')
+        smooth = self.smooth * len(df_sum)
+        df_sum[self.out_col] = (df_sum[f'{y_col}_x'] +
+                                smooth*self.mean) / \
+                               (df_sum[f'{y_col}_y'] +
+                                smooth)
+        return df_sum
+
+    def _groupby_agg(self, train, x_cols, op):
+        """
+        Compute aggregated value of each fold and overall dataframe
+        grouped by `x_cols` and agg by `op`
+        """
+        cols = [self.fold_col]+x_cols
+        df_each_fold = train.groupby(cols, as_index=False)\
+            .agg({self.y_col: op})
+        df_all = df_each_fold.groupby(x_cols, as_index=False)\
+            .agg({self.y_col: 'sum'})
+
+        df_each_fold = df_each_fold.merge(df_all, on=x_cols, how='left')
+        df_each_fold[f'{self.y_col}_x'] = df_each_fold[f'{self.y_col}_y'] -\
+            df_each_fold[f'{self.y_col}_x']
+        return df_each_fold, df_all
+
+    def _check_is_fitted(self):
+        if not self._fitted or self.train is None:
+            msg = ("This LabelEncoder instance is not fitted yet. Call 'fit' "
+                   "with appropriate arguments before using this estimator.")
+            raise NotFittedError(msg)
+
+    def _is_train_df(self, df):
+        """
+        Return True if the dataframe `df` is the training dataframe, which
+        is used in `fit_transform`
+        """
+        if len(df) != len(self.train):
+            return False
+        self.train = self.train.sort_values(self.id_col).reset_index(drop=True)
+        for col in df.columns:
+            if col not in self.train.columns:
+                raise ValueError(f"Input column {col} "
+                                 "is not in train data.")
+            if not (df[col] == self.train[col]).all():
+                return False
+        return True
+
+    def _impute_and_sort(self, df):
+        """
+        Impute and sort the result encoding in the same row order as input
+        """
+        df[self.out_col] = df[self.out_col].nans_to_nulls()
+        df[self.out_col] = df[self.out_col].fillna(self.mean)
+        df = df.sort_values(self.id_col)
+        res = df[self.out_col].values.copy()
+        if self.output_type == 'numpy':
+            return cp.asnumpy(res)
+        return res
+
+    def _to_cudf_frame(self, x):
```

Comment:
can be removed with my suggesion, renaming the columns afterwards. So perhaps the call of `input_to_cuml_array` could be done here and then just adding the column names in this method.